File: java/com/joshmanisdabomb/lcc/registry/LCCParticles.java
package com.joshmanisdabomb.lcc.registry;

import com.joshmanisdabomb.lcc.LCC;
import com.joshmanisdabomb.lcc.particle.HydratedSoulSandBubbleParticle;
import net.minecraft.client.Minecraft;
import net.minecraft.particles.BasicParticleType;
import net.minecraft.particles.ParticleType;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;
import net.minecraftforge.event.RegistryEvent;

import java.util.ArrayList;

public abstract class LCCParticles {

    public static final ArrayList<ParticleType<?>> all = new ArrayList<>();

    public static BasicParticleType hydrated_soul_sand_bubble;

    public static void init(RegistryEvent.Register<ParticleType<?>> e) {
        all.add(hydrated_soul_sand_bubble = (BasicParticleType)new BasicParticleType(false).setRegistryName(LCC.MODID, "hydrated_soul_sand_bubble"));
    }

    @OnlyIn(Dist.CLIENT)
    public static void initFactories() {
        Minecraft.getInstance().particles.registerFactory(hydrated_soul_sand_bubble, (type, world, x, y, z, xSpeed, ySpeed, zSpeed) -> new HydratedSoulSandBubbleParticle(world, x, y, z, xSpeed, ySpeed, zSpeed));
    }

}

File: kotlin/com/joshmanisdabomb/lcc/directory/LCCParticles.kt
package com.joshmanisdabomb.lcc.directory

import net.fabricmc.fabric.api.particle.v1.FabricParticleTypes
import net.minecraft.particle.ParticleType
import net.minecraft.util.registry.Registry

object LCCParticles : RegistryDirectory<ParticleType<*>, Unit>() {

    override val _registry by lazy { Registry.PARTICLE_TYPE }

    val soaking_soul_sand_bubble by create { FabricParticleTypes.simple(false) }
    val soaking_soul_sand_jump by create { FabricParticleTypes.simple(true) }

}